src/common/json-schema-input/array-input.js

```diff
@@ -0,0 +1,165 @@
+import _ from 'messages'
+import React, { Component, cloneElement } from 'react'
+import map from 'lodash/map'
+import filter from 'lodash/filter'
+
+import {
+  autobind,
+  propTypes
+} from 'utils'
+
+import GenericInput from './generic-input'
+import { descriptionRender } from './helpers'
+
+// ===================================================================
+
+class ArrayItem extends Component {
+  get value () {
+    return this.refs.input.value
+  }
+
+  set value (value) {
+    this.refs.input.value = value
+  }
+
+  render () {
+    const { props } = this
+
+    return (
+      <li className='list-group-item clearfix'>
+        {cloneElement(props.children, {
+          ref: 'input'
+        })}
+        <button className='btn btn-danger pull-xs-right' type='button' onClick={props.onDelete}>
+          {_('remove')}
+        </button>
+      </li>
+    )
+  }
+}
+
+// ===================================================================
+
+@propTypes({
+  depth: propTypes.number,
+  label: propTypes.string.isRequired,
+  required: propTypes.bool,
+  schema: propTypes.object.isRequired,
+  uiSchema: propTypes.object,
+  value: propTypes.object
+})
+export default class ArrayInput extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      use: props.required,
+      children: this._makeChildren(props)
+    }
+
+    this._nextChildKey = 0
+  }
+
+  get value () {
+    return map(this.refs, 'value')
+  }
+
+  @autobind
+  _handleOptionalChange (event) {
+    this.setState({
+      use: event.target.checked
+    })
+  }
+
+  @autobind
+  _handleAdd () {
+    const { children } = this.state
+    this.setState({
+      children: children.concat(this._makeChild(this.props))
+    })
+  }
+
+  _remove (key) {
+    this.setState({
+      children: filter(this.state.children, (child) => child.key !== key)
+    })
+  }
+
+  _makeChild (props) {
+    const key = String(this._nextChildKey++)
+    const {
+      schema: {
+        items
+      }
+    } = props
+
+    return (
+      <ArrayItem key={key} onDelete={() => { this._remove(key) }}>
+        <GenericInput
+          depth={props.depth}
+          label={items.title || _('item')}
+          required
+          schema={items}
+          uiSchema={props.uiSchema.items}
+        />
+      </ArrayItem>
+    )
+  }
+
+  _makeChildren ({ value, ...props }) {
+    return map(value, (value) => {
+      return (
+        this._makeChild({
+          ...props,
+          value
+        })
+      )
+    })
+  }
+
+  componentWillReceiveProps (props) {
+    this.setState({
```

Comment:
You should have test here if the props you are interested in have really changed.